mobilecoind-dev-faucet/src/lib.rs

```diff
@@ -0,0 +1,296 @@
+
+        // Generate a Tx sending this specific TxOut, less fees
+        let mut req = mc_mobilecoind_api::GenerateTxFromTxOutListRequest::new();
+        req.set_account_key((&self.account_key).into());
+        req.set_input_list(vec![utxo_record.utxo].into());
+        req.set_receiver(public_address.clone());
+        req.set_token_id(*token_id);
+
+        let resp = self
+            .mobilecoind_api_client
+            .generate_tx_from_tx_out_list_async(&req)
+            .map_err(|err| format!("Failed to build Tx: {}", err))?
+            .await
+            .map_err(|err| format!("Build Tx ended in error: {}", err))?;
+
+        // Submit the tx proposal
+        let mut req = mc_mobilecoind_api::SubmitTxRequest::new();
+        req.set_tx_proposal(resp.get_tx_proposal().clone());
+
+        let resp = self
+            .mobilecoind_api_client
+            .submit_tx_async(&req)
+            .map_err(|err| format!("Failed to submit Tx: {}", err))?
+            .await
+            .map_err(|err| format!("Submit Tx ended in error: {}", err))?;
+
+        // Tell the worker that this utxo was submitted, so that it can track and
+        // recycle the utxo if this payment fails
+        let _ = utxo_record.sender.send(resp.clone());
+        Ok(resp)
+    }
+
+    /// Handle a "get status" request to the faucet.
+    /// Returns either the json status report or an error string.
+    pub async fn handle_status(&self) -> Result<JsonFaucetStatus, String> {
+        // Get up-to-date balances for all the tokens we are tracking
+        let mut balances: HashMap<TokenId, u64> = Default::default();
+        for (token_id, _) in self.faucet_payout_amounts.iter() {
+            let mut req = mc_mobilecoind_api::GetBalanceRequest::new();
+            req.set_monitor_id(self.monitor_id.clone());
+            req.set_token_id(**token_id);
+
+            let resp = self
+                .mobilecoind_api_client
+                .get_balance_async(&req)
+                .map_err(|err| {
+                    format!(
+                        "Failed to check balance for token id '{}': {}",
+                        token_id, err
+                    )
+                })?
+                .await
+                .map_err(|err| {
+                    format!(
+                        "Balance check request for token id '{}' ended in error: {}",
+                        token_id, err
+                    )
+                })?;
+            balances.insert(*token_id, resp.balance);
+        }
+
+        let queue_depths = self.worker.get_queue_depths();
+
+        Ok(JsonFaucetStatus {
+            success: true,
+            err_str: None,
+            b58_address: self.monitor_b58_address.clone(),
+            faucet_payout_amounts: self
+                .faucet_payout_amounts
+                .iter()
+                .map(convert_balance_pair)
+                .collect(),
+            balances: balances.iter().map(convert_balance_pair).collect(),
```

Comment:
i tried this in an earlier revision, but `convert_balance_pair` only either works with references, or not with references. i figured having less code is more valuable than avoiding some copies, since we need it to work somewhere else with references.